File: nucleotide-count/nucleotide_count.ml
open Core.Std

(* Count the number of times the nucleotide occurs in the string. *)
let count dna n = String.count dna ~f:(fun c -> c = n)

(* Count the nucleotides in the string. *)
let nucleotide_counts dna =
  let inc m c = Char.Map.update m c ~f:(function
      | None ->  1
      | Some v -> v + 1) in
  String.fold dna ~init:Char.Map.empty ~f:inc

(* iterates the string 4 times
   let nucleotide_counts dna =
   ['A'; 'C'; 'G'; 'T']
   |> List.map ~f:(fun c -> (c, count dna c))
   |> List.filter ~f:(fun (_, v) -> v > 0)
   |> Char.Map.of_alist_exn
*)

File: fsharp/nucleotide-count/NucleotideCount.fs
module NucleotideCount

let nucleotideCounts (strand: string): Map<char, int> option =
  let empty = [('A',0); ('C',0); ('G', 0); ('T',0)] |> Map.ofList |> Some
  let count c (m:Map<char, int>) = match m.TryFind(c) with
                                   | Some v -> Some (m.Add(c, (v + 1)))
                                   | None -> None
  strand.ToCharArray () 
  |> Array.fold (fun a c -> match a with
                            | Some m -> count c m
                            | None -> None) empty
